### DISCUSSION TITLE
RCTDeviceEventEmitter bridgeless alternative (Turbo Models)

### DISCUSSION THREAD

**Discussion by DanielFrTB** [reactwg/react-native-new-architecture#213] [Libraries]
@cipolleschi 
Hi there,

I'm trying to send events from Turbo Models to the JS side. I understand that in the new architecture, EventDispatcher (https://github.com/react-native-menu/menu/pull/749/commits/247e0ec8b4f31c283c865dcd8d68519cbcfd4bb0) is necessary. However, it requires a viewTag, which isn't available in Turbo Modules.

Is there a bridgeless alternative to RCTDeviceEventEmitter for sending events from Turbo Models  (iOS and Android) ?

Thanks 

#144

**Comment by DanielFrTB** [comment]
I see that pass a Callback is not valid since it used bridge 


`

  public String addUnitToPageAndroid(String pageId, String mode, String placement, Callback callback) {
 });

package com.facebook.react.bridge;

public interface Callback {
    void invoke(Object... var1);
}
`

**Comment by cipolleschi** [comment]
Have a look at [this example](https://github.com/react-native-community/RNNewArchitectureLibraries/tree/feat/swift-event-emitter). It uses swift but shows how to send events from Native to JS for iOS

**Reply by DanielFrTB** [reply]
@cipolleschi Thanks,
I can see in this example that `RCTEventEmitter` internally  using ` @property (nonatomic, weak) RCTBridge *bridge;`

Does it's mean that RCTEventEmitter is going to be deprecated ?

**Reply by cipolleschi** [reply]
uhm... you are right. the example uses 0.71, when bridgeless was not there yet. 🤔 

Let me dig a little bit more into this.

**Comment by zerobertelprivat** [comment]
I am also totally interested in a solution to use events like implemented via RCTEventEmitter in the bridge mode. On JS Side i want to listen to a pure event, fired on the Native Side. For example, i need it to handle click interactions on Carplay Screen (init player and play music via RN Track Player). I found no fully understandable doc on the RN page. Thank you in advance :)

**Reply by cipolleschi** [reply]
@zerobertelprivat do you need to send an event in Bridge mode? why do you need a turbomodule?

For the time being, the best approach is:
* For the old architecture
    * in JS
    ```js
    import {NativeEventEmitter} from 'react-native';
    
    // when you want to register a listener
    new NativeEventEmitter<YourEventPayload>().addListener('youreventName', data => {
      // do stuff
    }) 
    ```
    * in native, your module needs to extend `RCTEventEmitter` and the Java/Kt EventEmitter
    * then call the `sendEventWithName` method

* For the New Architecture we have the [example](https://github.com/reactwg/react-native-new-architecture/blob/main/docs/turbo-modules.md#add-event-emitting-capabilities)

**Reply by zerobertelprivat** [reply]
Hi thanks for your answer :) -> I get emitting an event work for now.

In the example regarding the new Arch i am getting a TS Error ->

`import type {EventEmitter} from 'react-native/Libraries/Types/CodegenTypes';`
gives me
Module '"react-native/Libraries/Types/CodegenTypes"' has no exported member 'EventEmitter'.ts(2305)
Clicking on the import links me to the Codegen.d.ts file, which doesnt include this type.
Is it a bug or do i something wrong? I am on RN 0.76.1

Thanks nevertheless for your great work!

**Reply by cipolleschi** [reply]
Hum.. It should be there. This is the [0.76-branch with the EventEmitter](https://github.com/facebook/react-native/blob/a9a1c86a927fc6e3854a9b4ad44d38bd3c8db588/packages/react-native/types/modules/Codegen.d.ts#L77-L79) definition.

Probably the import path is wrong. It should be `'react-native/types/modules/Codegen'`.

**Reply by agoransson** [reply]
If upgrading to 0.76.3 the import seems to work.

`import type {EventEmitter} from 'react-native/Libraries/Types/CodegenTypes';`
